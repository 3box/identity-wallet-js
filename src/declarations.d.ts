// not sure why this is needed
declare module 'ipfs' {
  export default any
}

declare module 'multihashes' {
  export function encode(digest: Buffer, code: number | string, length?: number): Buffer
}
