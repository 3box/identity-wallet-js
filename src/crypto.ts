import { randomBytes } from '@stablelib/random'
export { randomBytes }
import {
  createJWE,
  decryptJWE,
  JWE,
  Encrypter,
  Decrypter,
  x25519Encrypter,
  x25519Decrypter,
} from 'did-jwt'
import { prepareCleartext, decodeCleartext } from 'dag-jose-utils'

interface EncryptOpts {
  publicKey?: Uint8Array
  encrypter?: Encrypter
  kid?: string
}

function isU8a(b: Uint8Array | undefined): b is Uint8Array {
  return b instanceof Uint8Array
}

export async function asymEncryptJWE(
  cleartext: Record<string, any>,
  { publicKey, encrypter, kid }: EncryptOpts
): Promise<JWE> {
  if (!encrypter) {
    if (!isU8a(publicKey)) throw new Error('publicKey or encrypter has to be defined')
    encrypter = x25519Encrypter(publicKey, kid)
  }
  return createJWE(prepareCleartext(cleartext), [encrypter])
}

interface DecryptOpts {
  secretKey?: Uint8Array
  decrypter?: Decrypter
}

export async function asymDecryptJWE(
  jwe: JWE,
  { secretKey, decrypter }: DecryptOpts
): Promise<Record<string, any>> {
  if (!decrypter) {
    if (!isU8a(secretKey)) throw new Error('secretKey or decrypter has to be defined')
    decrypter = x25519Decrypter(secretKey)
  }
  return decodeCleartext(await decryptJWE(jwe, decrypter))
}

interface Recipient {
  header: Record<string, string>
}

export function parseJWEKids(jwe: JWE): Array<string> {
  return (jwe.recipients || []).reduce((kids: Array<string>, recipient: Recipient) => {
    if (recipient.header?.kid) kids.push(recipient.header.kid.split('#')[1])
    return kids
  }, [])
}
