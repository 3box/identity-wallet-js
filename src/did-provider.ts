import { createJWS } from 'did-jwt'
import {
  HandlerMethods,
  RequestHandler,
  RPCConnection,
  RPCError,
  RPCRequest,
  RPCResponse,
  createHandler,
} from 'rpc-utils'

import Keyring from './keyring'
import { ThreeIDX } from './three-idx'
import Permissions from './permissions'
import { toStableObject } from './utils'

type Origin = string | null | undefined

export type Context = {
  permissions: Permissions
  threeIdx: ThreeIDX
  keyring: Keyring
  origin: Origin
}

interface CreateJWSParams {
  payload: Record<string, any>
  protected?: Record<string, any>
  did: string
}

interface AuthParams {
  paths: Array<string>
}

export const didMethods: HandlerMethods<Context> = {
  did_authenticate: async ({ permissions, threeIdx, origin }, params: AuthParams) => {
    const paths = await permissions.request(origin, params.paths || [])
    // paths should be an array if permission granted
    // may be a subset or requested paths or empty array
    if (paths === null) throw new RPCError(4001, 'User Rejected Request')
    return { did: threeIdx.DID, paths }
  },
  did_createJWS: async ({ permissions, keyring, threeIdx, origin }, params: CreateJWSParams) => {
    if (!permissions.has(origin)) {
      throw new RPCError(4100, 'Unauthorized')
    }
    // TODO - if the requesting DID is our management key
    // (did:key) we should request explicit permission.
    const keyName = threeIdx.parseKeyName(params.did)
    const kid = await threeIdx.encodeKidWithVersion(keyName)
    const signer = keyring.getSigner(keyName)
    const header = toStableObject(Object.assign(params.protected || {}, { kid }))
    const jws = await createJWS(toStableObject(params.payload), signer, header)
    return { jws }
  },
}

export interface ProviderConfig {
  permissions: Permissions
  threeIdx: ThreeIDX
  keyring: Keyring
  forcedOrigin?: string
}

export class DidProvider implements RPCConnection {
  protected _handle: RequestHandler

  constructor({ permissions, threeIdx, keyring, forcedOrigin }: ProviderConfig) {
    const handler = createHandler<Context>(didMethods)
    this._handle = (origin: string, msg: RPCRequest) => {
      return handler(
        {
          origin: forcedOrigin || origin,
          permissions,
          threeIdx,
          keyring,
        },
        msg
      )
    }
  }

  public get isDidProvider(): boolean {
    return true
  }

  public async send(msg: RPCRequest, origin?: Origin): Promise<RPCResponse | null> {
    return await this._handle(origin, msg)
  }
}
