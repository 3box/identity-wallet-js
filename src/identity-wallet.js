import EventEmitter from 'events'
import Keyring from './keyring'
import didJWT from 'did-jwt'
import { sha256Multihash, pad, unpad } from './utils'


class IdentityWallet {

  /**
   * Creates an instance of IdentityWallet
   *
   * @param     {Object}    config                  The configuration to be used
   * @param     {String}    config.seed             The seed of the identity, 32 hex string
   * @param     {String}    config.authSecret       The authSecret to use, 32 hex string
   * @param     {String}    config.ethereumAddress  The ethereumAddress of the identity
   * @return    {this}                            An IdentityWallet instance
   */
  constructor (config = {}) {
    this.events = new EventEmitter()
    if (config.seed) {
      this._keyring = new Keyring(config.seed)
    }
    if (config.authSecret) {
      if (!config.ethereumAddress) throw new Error('Ethereum address needs to be defined when authSecret given')
      this._authSecret = config.authSecret
      this._ethereumAddress = config.ethereumAddress
    }
  }

  async getAddress () {
    return this._ethereumAddress
  }

  _initKeyring (authData) {
    let seed
    if (authData) {
      authData.find(({ ciphertext, nonce }) => {
        const key = Keyring.hexToUint8Array(this._authSecret)
        seed = Keyring.symDecryptBase(ciphertext, key, nonce)
        return Boolean(seed)
      })
      if (!seed) throw new Error('No valid auth-secret for this identity')
    } else {
      // no authData available so we create a new identity
      seed = '0x' + Buffer.from(Keyring.naclRandom(32)).toString('hex')
      this.addAuthMethod(this._authSecret, seed)
    }
    this._keyring = new Keyring(seed)
  }

  /**
   * Authenticate to given spaces
   *
   * @param     {Array<String>}     spaces          The desired spaces
   * @param     {Object}    opts                    Optional parameters
   * @param     {Array<Object>}     opts.authData   The authData for this identity
   * @return    {Object}                            The public keys for the requested spaces of this identity
   */
  async authenticate (spaces = [], { authData } = {}) {
    if (!this._keyring) this._initKeyring(authData)
    const result = {
      main: this._keyring.getPublicKeys(),
      spaces: {}
    }
    if (this._ethereumAddress) result.main.managementKey = this._ethereumAddress
    spaces.map(space => {
      result.spaces[space] = this._keyring.getPublicKeys({ space, uncompressed: true })
    })
    return result
  }

  /**
   * Check if authenticated to given spaces
   *
   * @param     {Array<String>}     spaces          The desired spaces
   * @return    {Boolean}                           True if authenticated
   */
  async isAuthenticated (spaces = []) {
    return Boolean(this._keyring)
  }

  /**
   * Add a new authentication method for this identity
   *
   * @param     {String}    authSecret              A 32 byte hex string used as authentication secret
   */
  async addAuthMethod (authSecret, seed) {
    if (!seed && !this._keyring) throw new Error('This method can only be called after authenticate has been called')

    const message = seed || this._keyring._seed
    const key = Keyring.hexToUint8Array(authSecret)
    const encAuthData = Keyring.symEncryptBase(message, key)

    this.events.emit('new-auth-method', encAuthData)
  }

  /**
   * Sign a verifiable credential. The format of the credential is [did-jwt](https://github.com/uport-project/did-jwt).
   *
   * @param     {Object}    payload                 The payload of the claim
   * @param     {Object}    opts                    Optional parameters
   * @param     {String}    opts.DID                The DID used as the issuer of this claim
   * @param     {String}    opts.space              The space used to sign the claim
   * @param     {String}    opts.expiresIn          Set an expiry date for the claim as unix timestamp
   * @return    {String}                            The signed claim encoded as a JWT
   */
  async signClaim (payload, { DID, space, expiresIn } = {}) {
    if (!this._keyring) throw new Error('This method can only be called after authenticate has been called')

    const issuer = DID
    const settings = {
      signer: this._keyring.getJWTSigner(space),
      issuer,
      expiresIn
    }
    return didJWT.createJWT(payload, settings)
  }

  /**
   * Encrypt a message
   *
   * @param     {String}    message                 The message to be encrypted
   * @param     {String}    space                   The space used for encryption
   * @param     {Object}    opts                    Optional parameters
   * @param     {String}    opts.nonce              The nonce used to encrypt the message
   * @param     {String}    opts.blockSize          The blockSize used for padding (default 24)
   * @return    {Object}                            The encrypted object (ciphertext and nonce)
   */
  async encrypt (message, space, { nonce, blockSize } = {}) {
    if (!this._keyring) throw new Error('This method can only be called after authenticate has been called')

    const paddedMsg = pad(message, blockSize)
    return this._keyring.symEncrypt(paddedMsg, { space, nonce })
  }

  /**
   * Decrypt a message
   *
   * @param     {Object}    encryptedObject         The encrypted object (ciphertext and nonce)
   * @param     {String}    space                   The space used for encryption
   * @return    {String}                            The decrypted message
   */
  async decrypt (encObj, space) {
    if (!this._keyring) throw new Error('This method can only be called after authenticate has been called')

    const paddedMsg = this._keyring.symDecrypt(encObj.ciphertext, encObj.nonce, { space })
    if (!paddedMsg) throw new Error('IdentityWallet: Could not decrypt message')
    return unpad(paddedMsg)
  }

  async hashDBKey(key, space) {
    const salt = this._keyring.getDBSalt(space)
    return sha256Multihash(salt + key)
  }
}


module.exports = IdentityWallet
