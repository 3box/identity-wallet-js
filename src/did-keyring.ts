import nacl, { BoxKeyPair } from 'tweetnacl'
import naclutil from 'tweetnacl-util'
import { HDNode } from '@ethersproject/hdnode'
import { Wallet } from '@ethersproject/wallet'
import { SimpleSigner } from 'did-jwt'
import { sha256 } from 'js-sha256'
import { ec as EC } from 'elliptic'

import { naclRandom } from './crypto'

const ec = new EC('secp256k1')

const BASE_PATH = "m/51073068'/0'"
const ROOT_STORE_PATH = "0'/0'/0'/0'/0'/0'/0'/0'"

const AUTH_PATH_WALLET = BASE_PATH + '/' + ROOT_STORE_PATH + '/0'
const AUTH_PATH_ENCRYPTION = BASE_PATH + '/' + ROOT_STORE_PATH + '/3'

const ensure0x = (str: string): string => {
  return (str.startsWith('0x') ? '' : '0x') + str
}

interface KeysSet {
  signingKey: HDNode
  asymEncryptionKey: BoxKeyPair
  symEncryptionKey: Uint8Array
  managementKey: HDNode
}

export class Keyring {
  public static create() {
    const seed = '0x' + Buffer.from(naclRandom(32)).toString('hex')
    return new Keyring(seed)
  }

  private _seed: string
  private _baseNode: HDNode
  private _keys: KeysSet

  constructor(seed: string) {
    this._seed = seed
    this._baseNode = HDNode.fromSeed(seed).derivePath(BASE_PATH)
    const rootNode = this._baseNode.derivePath(ROOT_STORE_PATH)
    this._keys = {
      signingKey: rootNode.derivePath('0'),
      asymEncryptionKey: nacl.box.keyPair.fromSecretKey(
        new Uint8Array(
          Buffer.from(rootNode.derivePath('2').privateKey.slice(2), 'hex'),
        ),
      ),
      symEncryptionKey: Keyring.hexToUint8Array(
        rootNode.derivePath('3').privateKey.slice(2),
      ),
      managementKey: rootNode.derivePath('1'),
    }
  }

  _getKeys() {
    return this._keys
  }

  asymEncrypt(msg, toPublic, { nonce } = {}) {
    nonce = nonce || Keyring.randomNonce()
    toPublic = naclutil.decodeBase64(toPublic)
    if (typeof msg === 'string') {
      msg = naclutil.decodeUTF8(msg)
    }
    const ephemneralKeypair = nacl.box.keyPair()
    const ciphertext = nacl.box(
      msg,
      nonce,
      toPublic,
      ephemneralKeypair.secretKey,
    )
    return {
      nonce: naclutil.encodeBase64(nonce),
      ephemeralFrom: naclutil.encodeBase64(ephemneralKeypair.publicKey),
      ciphertext: naclutil.encodeBase64(ciphertext),
    }
  }

  asymDecrypt(ciphertext, fromPublic, nonce, { space, toBuffer } = {}) {
    fromPublic = naclutil.decodeBase64(fromPublic)
    ciphertext = naclutil.decodeBase64(ciphertext)
    nonce = naclutil.decodeBase64(nonce)
    const cleartext = nacl.box.open(
      ciphertext,
      nonce,
      fromPublic,
      this._keys.asymEncryptionKey.secretKey,
    )
    if (toBuffer) {
      return cleartext ? Buffer.from(cleartext) : null
    }
    return cleartext ? nacl.util.encodeUTF8(cleartext) : null
  }

  symEncrypt(msg, { space, nonce } = {}) {
    return Keyring.symEncryptBase(msg, this._keys.symEncryptionKey, nonce)
  }

  symDecrypt(ciphertext, nonce, { space, toBuffer } = {}) {
    return Keyring.symDecryptBase(
      ciphertext,
      this._keys.symEncryptionKey,
      nonce,
      toBuffer,
    )
  }

  managementPersonalSign(message) {
    const wallet = this.managementWallet()
    return wallet.signMessage(message)
  }

  managementWallet() {
    return new Wallet(this._keys.managementKey.privateKey)
  }

  getJWTSigner() {
    return SimpleSigner(this._keys.signingKey.privateKey.slice(2))
  }

  getDBSalt() {
    return sha256(this._keys.signingKey.derivePath('0').privateKey.slice(2))
  }

  getPublicKeys({ uncompressed }) {
    let signingKey = this._keys.signingKey.publicKey.slice(2)
    const managementKey = this._keys.managementKey.address
    if (uncompressed) {
      signingKey = ec
        .keyFromPublic(Buffer.from(signingKey, 'hex'))
        .getPublic(false, 'hex')
    }
    return {
      signingKey,
      managementKey,
      asymEncryptionKey: naclutil.encodeBase64(
        this._keys.asymEncryptionKey.publicKey,
      ),
    }
  }

  serialize() {
    return this._seed
  }

  static encryptWithAuthSecret(message, authSecret) {
    const node = HDNode.fromSeed(ensure0x(authSecret)).derivePath(
      AUTH_PATH_ENCRYPTION,
    )
    const key = Keyring.hexToUint8Array(node.privateKey.slice(2))
    return Keyring.symEncryptBase(message, key)
  }

  static decryptWithAuthSecret(ciphertext, nonce, authSecret) {
    const node = HDNode.fromSeed(ensure0x(authSecret)).derivePath(
      AUTH_PATH_ENCRYPTION,
    )
    const key = Keyring.hexToUint8Array(node.privateKey.slice(2))
    return Keyring.symDecryptBase(ciphertext, key, nonce)
  }

  static walletForAuthSecret(authSecret) {
    const node = HDNode.fromSeed(ensure0x(authSecret)).derivePath(
      AUTH_PATH_WALLET,
    )
    return new Wallet(node.privateKey)
  }

  static hexToUint8Array(str) {
    return new Uint8Array(Buffer.from(str, 'hex'))
  }

  static symEncryptBase(msg, symKey, nonce) {
    nonce = nonce || Keyring.randomNonce()
    if (typeof msg === 'string') {
      msg = nacl.util.decodeUTF8(msg)
    }
    const ciphertext = nacl.secretbox(msg, nonce, symKey)
    return {
      nonce: nacl.util.encodeBase64(nonce),
      ciphertext: nacl.util.encodeBase64(ciphertext),
    }
  }

  static symDecryptBase(ciphertext, symKey, nonce, toBuffer) {
    ciphertext = nacl.util.decodeBase64(ciphertext)
    nonce = nacl.util.decodeBase64(nonce)
    const cleartext = nacl.secretbox.open(ciphertext, nonce, symKey)
    if (toBuffer) {
      return cleartext ? Buffer.from(cleartext) : null
    }
    return cleartext ? nacl.util.encodeUTF8(cleartext) : null
  }

  static naclRandom(length) {
    return nacl.randomBytes(length)
  }

  static randomNonce() {
    return Keyring.naclRandom(24)
  }
}

module.exports = Keyring
