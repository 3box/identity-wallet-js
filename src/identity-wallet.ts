import { CeramicApi } from '@ceramicnetwork/ceramic-common'

import { DidProvider } from './did-provider'
import Keyring from './keyring'
import ThreeIdProvider from './threeIdProvider'
import { ThreeIDX } from './three-idx'
import Permissions, { GetPermissionFn, SELF_ORIGIN } from './permissions'
import { Keychain } from './keychain'

interface IDWConfig {
  getPermission: GetPermissionFn
  seed?: string
  authSecret?: Uint8Array
  authId?: string
  externalAuth?: (req: any) => Promise<any>
  ceramic: CeramicApi
  useThreeIdProv: boolean
}

export default class IdentityWallet {
  protected _externalAuth: ((req: any) => Promise<any>) | undefined

  /**
   * Use IdentityWallet.create() to create an IdentityWallet instance
   */
  constructor(
    protected _keyring: Keyring,
    protected _threeIdx: ThreeIDX,
    protected _permissions: Permissions,
    protected _keychain: Keychain
  ) {}
  /**
   * @property {Keychain} keychain          Edit the keychain
   */
  get keychain(): Keychain {
    return this._keychain
  }

  /**
   * @property {Permissions} permissions    Edit permissions
   */
  get permissions(): Permissions {
    return this._permissions
  }

  /**
   * @property {string} DID                 The 3ID of the IdentityWallet instance
   */
  get DID(): string {
    return this._threeIdx.DID
  }

  /**
   * Creates an instance of IdentityWallet
   *
   * @param     {Object}        config                  The configuration to be used
   * @param     {Function}      config.getPermission    The function that is called to ask the user for permission
   * @param     {String}        config.seed             The seed of the identity, 32 bytes hex string
   * @param     {Uint8Array}    config.authSecret       The authSecret to use, 32 bytes
   * @param     {String}        config.authId           The authId is used to identify the authSecret
   * @param     {String}        config.externalAuth     External auth function, directly returns key material, used to migrate legacy 3box accounts
   * @return    {IdentityWallet}                        An IdentityWallet instance
   */
  static async create(config: IDWConfig): Promise<IdentityWallet> {
    if (config.seed && config.authSecret) throw new Error("Can't use both seed and authSecret")
    if (!config.seed && !config.authSecret) throw new Error('Either seed or authSecret is needed')
    if (config.authSecret && !config.authId) {
      throw new Error('AuthId must be given along with authSecret')
    }
    const threeIdx = new ThreeIDX(config.ceramic)
    const permissions = new Permissions(config.getPermission)
    let keyring, keychain
    if (config.seed) {
      keyring = new Keyring(config.seed)
      keychain = new Keychain(keyring, threeIdx)
      const pubkeys = keyring.getPublicKeys({ mgmtPub: true, useMulticodec: true })
      await threeIdx.create3idDoc(pubkeys)
    } else if (config.authSecret) {
      keychain = await Keychain.load(threeIdx, config.authSecret)
      keyring = keychain._keyring
    }
    permissions.setDID(threeIdx.DID)
    const idw = new IdentityWallet(keyring as Keyring, threeIdx, permissions, keychain as Keychain)
    await idw._threeIdx.setDIDProvider(idw.getDidProvider(SELF_ORIGIN))
    if (config.authId && !keychain?.list().length) {
      // Add the auth method to the keychain
      await idw.keychain.add(config.authId, config.authSecret as Uint8Array)
      await idw.keychain.commit()
    }
    return idw
  }

  /**
   * Get the DIDProvider
   *
   * @return    {DidProvider}                   The DIDProvider for this IdentityWallet instance
   */
  getDidProvider(forcedOrigin?: string): DidProvider {
    return new DidProvider({
      keyring: this._keyring,
      permissions: this.permissions,
      threeIdx: this._threeIdx,
      forcedOrigin,
    })
  }

  /**
   * Get the 3IDProvider
   *
   * @return    {ThreeIdProvider}                   The 3IDProvider for this IdentityWallet instance
   */
  get3idProvider(forcedOrigin?: string): ThreeIdProvider {
    return new ThreeIdProvider({
      keyring: this._keyring,
      permissions: this.permissions,
      threeIdx: this._threeIdx,
      forcedOrigin,
    })
  }
}
