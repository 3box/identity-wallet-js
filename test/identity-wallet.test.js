import IdentityWallet from '../src/identity-wallet'
import { randomBytes } from '../src/crypto'

import { verifyJWT } from 'did-jwt'
import { Resolver } from 'did-resolver'
import tmp from 'tmp-promise'
import Ceramic from '@ceramicnetwork/ceramic-core'
import Ipfs from 'ipfs'
import { publishIDXConfig } from '@ceramicstudio/idx-tools'

import dagJose from 'dag-jose'
import basicsImport from 'multiformats/cjs/src/basics-import.js'
import legacy from 'multiformats/cjs/src/legacy.js'

const seed = '0x6e34b2e1a9624113d81ece8a8a22e6e97f0e145c25c1d4d2d0e62753b4060c837097f768559e17ec89ee20cba153b23b9987912ec1e860fa1212ba4b84c776ce'

const randomAuthSecret = () => randomBytes(32)
const getPermissionMock = jest.fn(async () => [])

const genIpfsConf = (folder) => {
  basicsImport.multicodec.add(dagJose)
  const format = legacy(basicsImport, dagJose.name)
  return {
    ipld: { formats: [format] },
    repo: `${folder}/ipfs/`,
    config: {
      Addresses: { Swarm: [] },
      Bootstrap: []
    },
  }
}

describe('IdentityWallet', () => {
  jest.setTimeout(15000)
  let tmpFolder
  let ipfs, ceramic

  beforeAll(async () => {
    tmpFolder = await tmp.dir({ unsafeCleanup: true })
    ipfs = await Ipfs.create(genIpfsConf(tmpFolder.path))
    ceramic = await Ceramic.create(ipfs, { stateStorePath: tmpFolder.path + '/ceramic/'})
    await publishIDXConfig(ceramic)
  })

  afterAll(async () => {
    await ceramic.close()
    await ipfs.stop()
    await tmpFolder.cleanup()
  })

  describe('.create', () => {
    it('Creates instance from seed', async () => {
      const config = {
        getPermission: getPermissionMock,
        seed,
        ceramic
      }
      const idw = await IdentityWallet.create(config)
      expect(await ceramic.context.resolver.resolve(idw.id)).toBeDefined()
      expect(idw.keychain.list()).toEqual([])
    })

    it('Creates instance from authSecret, new DID', async () => {
      const config = {
        getPermission: getPermissionMock,
        authSecret: randomAuthSecret(),
        authId: 'testAuth',
        ceramic
      }
      const idw = await IdentityWallet.create(config)
      expect(await ceramic.context.resolver.resolve(idw.id)).toBeDefined()
      expect(idw.keychain.list()).toEqual(['testAuth'])
    })

    it('Creates instance from authSecret, existing DID', async () => {
      const config = {
        getPermission: getPermissionMock,
        authSecret: randomAuthSecret(),
        authId: 'testAuth',
        ceramic
      }
      const idw1 = await IdentityWallet.create(config)
      expect(await ceramic.context.resolver.resolve(idw1.id)).toBeDefined()
      expect(idw1.keychain.list()).toEqual(['testAuth'])

      const idw2 = await IdentityWallet.create(config)
      expect(idw2.keychain.list()).toEqual(['testAuth'])
      expect(idw1.id).toEqual(idw2.id)
    })
  })

  describe('.keychain', () => {
    it('Adds authSecret to the keychain', async () => {
      const config1 = { getPermission: getPermissionMock, seed, ceramic }
      const idw1 = await IdentityWallet.create(config1)
      expect(idw1.keychain.list()).toEqual([])

      await idw1.keychain.add('auth2', randomAuthSecret())
      await idw1.keychain.commit()
      expect(idw1.keychain.list()).toEqual(['auth2'])
    })

    it('Creates instance from added authSecret', async () => {
      const config1 = { getPermission: getPermissionMock, seed, ceramic }
      const idw1 = await IdentityWallet.create(config1)
      expect(idw1.keychain.list()).toEqual([])

      const config2 = {
        getPermission: getPermissionMock,
        authId: 'auth2',
        authSecret: randomAuthSecret(),
        ceramic
      }
      await idw1.keychain.add(config2.authId, config2.authSecret)
      await idw1.keychain.commit()

      const idw2 = await IdentityWallet.create(config2)
      expect(idw2.keychain.list()).toEqual(['auth2'])
      expect(idw1.id).toEqual(idw2.id)
    })
  })
})

describe('IdentityWallet with disabled IDX', () => {
  jest.setTimeout(15000)
  let tmpFolder
  let ipfs, ceramic

  beforeAll(async () => {
    tmpFolder = await tmp.dir({ unsafeCleanup: true })
    ipfs = await Ipfs.create(genIpfsConf(tmpFolder.path))
    ceramic = await Ceramic.create(ipfs, { stateStorePath: tmpFolder.path + '/ceramic/' })
  })

  afterAll(async () => {
    await ceramic.close()
    await ipfs.stop()
    await tmpFolder.cleanup()
  })

  describe('.create', () => {
    it('Creates instance from seed', async () => {
      const config = {
        getPermission: getPermissionMock,
        seed,
        ceramic,
        disableIDX: true,
      }
      const idw = await IdentityWallet.create(config)
      expect(await ceramic.context.resolver.resolve(idw.id)).toBeDefined()
      expect(idw.keychain.list()).toEqual([])
      expect(idw._threeIdx.docs.idx).toBeUndefined()
    })

    it('Throws when trying to creates instance from authSecret', async () => {
      const config = {
        getPermission: getPermissionMock,
        authSecret: randomAuthSecret(),
        authId: 'testAuth',
        ceramic,
        disableIDX: true,
      }
      await expect(IdentityWallet.create(config)).rejects.toThrow(
        'AuthId cannot be used with disableIDX'
      )
    })
  })
})
