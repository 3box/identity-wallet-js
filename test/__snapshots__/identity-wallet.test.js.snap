// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IdentityWallet .create Creates instance from seed 1`] = `
Object {
  "@context": "https://w3id.org/did/v1",
  "authentication": Array [
    Object {
      "publicKey": "did:3:bafyreiaaskoyjlkq7a5mwdum5pzvhn3r45lqsmp7v622yko3etqc4wpmje#signing",
      "type": "Secp256k1SignatureAuthentication2018",
    },
  ],
  "id": "did:3:bafyreiaaskoyjlkq7a5mwdum5pzvhn3r45lqsmp7v622yko3etqc4wpmje",
  "keyAgreement": Array [
    Object {
      "controller": "did:3:bafyreiaaskoyjlkq7a5mwdum5pzvhn3r45lqsmp7v622yko3etqc4wpmje",
      "id": "did:3:bafyreiaaskoyjlkq7a5mwdum5pzvhn3r45lqsmp7v622yko3etqc4wpmje#encryption",
      "publicKeyBase58": "AVgughn1R4BTrRL2gP6ZmVAhWrzTHRx9jjbi5iKkeXK6",
      "type": "X25519KeyAgreementKey2019",
    },
  ],
  "publicKey": Array [
    Object {
      "controller": "did:3:bafyreiaaskoyjlkq7a5mwdum5pzvhn3r45lqsmp7v622yko3etqc4wpmje",
      "id": "did:3:bafyreiaaskoyjlkq7a5mwdum5pzvhn3r45lqsmp7v622yko3etqc4wpmje#signing",
      "publicKeyHex": "027ab5238257532f486cbeeac59a5721bbfec2f13c3d26516ca9d4c5f0ec1aa229",
      "type": "Secp256k1VerificationKey2018",
    },
    Object {
      "controller": "did:3:bafyreiaaskoyjlkq7a5mwdum5pzvhn3r45lqsmp7v622yko3etqc4wpmje",
      "id": "did:3:bafyreiaaskoyjlkq7a5mwdum5pzvhn3r45lqsmp7v622yko3etqc4wpmje#encryption",
      "publicKeyBase64": "jRKRy3oPhXpUBrFJefO2CJnWw/IoalOPuepAwvTqrEk=",
      "type": "Curve25519EncryptionPublicKey",
    },
  ],
  "service": Array [
    Object {
      "id": "did:3:bafyreiaaskoyjlkq7a5mwdum5pzvhn3r45lqsmp7v622yko3etqc4wpmje#idx",
      "serviceEndpoint": "ceramic://bagcqceralak3hhzv5xom27djnpq5mi7fgozn6ssar4yhhjfq4qcax2bfemnq",
      "type": "IdentityIndexRoot",
    },
  ],
}
`;
